pina/geometry/difference_domain.py

```diff
@@ -1,28 +1,142 @@
+        """ PINA implementation of Difference of Domains.
 
-        self.first = first
-        self.second = second
+        :param list geometries: A list of geometries from 'pina.geometry' 
+            such as 'EllipsoidDomain' or 'CartesianDomain'.
 
-    def sample(self, n, mode='random', variables='all'):
+        :Example:
+            # Create two ellipsoid domains
+            >>> ellipsoid1 = EllipsoidDomain({'x': [-1, 1], 'y': [-1, 1]})
+            >>> ellipsoid2 = EllipsoidDomain({'x': [0, 2], 'y': [0, 2]})
+
+            # Create a difference of the ellipsoid domains
+            >>> difference = Difference([ellipsoid1, ellipsoid2])
         """
+        super().__init__()
+
+        # difference checks
+        check_consistency(geometries, Location)
+        self._check_difference_dimesions(geometries)
+
+        # assign geometries
+        self._geometries = geometries
+
+    @property
+    def geometries(self):
+        """ 
+        The geometries."""
+        return self._geometries
+
+    @property
+    def variables(self):
         """
-        assert mode == 'random', 'Only random mode is implemented'
+        Spatial variables.
 
-        samples = []
-        while len(samples) < n:
-            sample = self.first.sample(1, 'random')
-            if not self.second.is_inside(sample):
-                samples.append(sample)
+        :return: All the spatial variables defined in '__init__()' in order.
+        :rtype: list[str]
+        """
+        all_variables = []
+        seen_variables = set()
+        for geometry in self.geometries:
+            for variable in geometry.variables:
+                if variable not in seen_variables:
+                    all_variables.append(variable)
+                    seen_variables.add(variable)
+        return all_variables
+
+    def is_inside(self, point, check_border=False):
+        """Check if a point is inside the difference domain.
+
+        :param point: Point to be checked.
+        :type point: torch.Tensor   
+        :param bool check_border: If True, the border is considered inside.
+        :return: True if the point is inside the difference domain, False otherwise.
+        :rtype: bool
+        """
+        flag = 0
+        for geometry in self.geometries:
+            if geometry.is_inside(point, check_border):
+                flag += 1
+        return flag == 1
+
+    def sample(self, n, mode='random', variables='all'):
+        """Sample routine.
 
-        import torch
-        return LabelTensor(torch.cat(samples), labels=['x', 'y'])
+        :param n: Number of points to sample in the shape.
+        :type n: int
+        :param mode: Mode for sampling, defaults to 'random'.
+            Available modes include: random sampling, 'random'.
+        :type mode: str, optional
+        :param variables: pinn variable to be sampled, defaults to 'all'.
+        :type variables: str or list[str], optional
+
+        :Example:
```

Comment:
Are you sure about this example? The LabelTensor you obtain doesn't seem to be correct. How can you  obtain `2.1999` on the `x` axis if you are doing `[0,2] - [1,3] = [0, 1]`?

Reply:
I went ahead and plotted the difference of the two CartesianDomains given in this example and the point (2.1999, 2.0767):
```
cart1 = CartesianDomain({'x': [0, 2], 'y': [0, 2]})
cart2 = CartesianDomain({'x': [1, 3], 'y': [1, 3]})
pt1 = LabelTensor(torch.tensor([[2.1999, 2.0767]]), ['x', 'y'])
difference = Difference([cart1, cart2])
```
The difference is in red and the point is in blue (enlarged to show a little bit more clearly). It looks like it's still in. I might be misinterpreting something so please let me know and I change it.

<img width="1154" alt="Screenshot 2023-06-21 at 11 50 37 AM" src="https://github.com/mathLab/PINA/assets/54781152/a87bc8d9-fdf6-4735-a8cd-e207ac93479e">

Reply:
I expected something more similar to this:

![Figure_1](https://github.com/mathLab/PINA/assets/93731561/8d5522d7-513f-4a49-b652-b71765467474)

@ndem0 What do you think?